RelhaxModpack/RelhaxModpack/Atlases/Packing/ImagePacker.cs

```diff
@@ -152,20 +168,69 @@ public class ImagePacker
             {
-                Rectangle location = imagePlacement[texture];
+                CopyTextureIntoAtlasLock(ref atlasByte, ref texture.AtlasImage, imagePlacement[texture], atlasData.Stride);
+            }
+
+            //copy back and unlock
+            Marshal.Copy(atlasByte, 0, atlasData.Scan0, atlasByteDepth);
+            atlas.UnlockBits(atlasData);
 
-                // copy pixels over to avoid anti-aliasing or any other side effects of drawing
-                // the sub-images to the output image using Graphics
-                for (int x = 0; x < texture.AtlasImage.Width; x++)
-                    for (int y = 0; y < texture.AtlasImage.Height; y++)
-                        outputImage.SetPixel(location.X + x, location.Y + y, texture.AtlasImage.GetPixel(x, y));
+            //for debugging, verify the image is bit for bit accurate
+            //https://online-image-comparison.com/
+            //http://onlinemd5.com/
+            //atlas.Save(@"C:\Users\Willster419\Desktop\custom2.png",ImageFormat.Png);
+
+            return atlas;
+        }
+
+        private void CopyTextureIntoAtlasLock(ref byte[] atlasByte, ref Bitmap texture, Rectangle locationOnAtlas, int atlasStride)
+        {
+            //define the area of the atlas that we actually want to copy over (don't copy padding)
+            Rectangle actualLocationToCopyOntoAtlas = new Rectangle(locationOnAtlas.X, locationOnAtlas.Y, texture.Width, texture.Height);
+
+            //lock the texture and get type data
+            BitmapData textureData = texture.LockBits(new Rectangle(0, 0, texture.Width, texture.Height), ImageLockMode.ReadOnly, texture.PixelFormat);
+            int textureByteDepth = Math.Abs(textureData.Stride) * actualLocationToCopyOntoAtlas.Height;
+            byte[] textureByte = new byte[textureByteDepth];
+            Marshal.Copy(textureData.Scan0, textureByte, 0, textureByteDepth);
+
+            /*
+             * 4 places in the array = 1 pixel. the alpha, red, green, blue
+             * so to get pixel 3 (1 based), it would be byte indexes 8 to 11
+             * the formula here is base ((pixel-1) * 4) to ((pixel-1) * 4) + 4
+             * 
+             * the amount that you want to copy for each row is width * 4
+             * and will loop for texture height
+             * 
+             * to get the starting point, multiply stride * (Y + loop), then X * 4
+            */
+
+            int tempTextureCount = 0;
+            for (int row = 0; row < actualLocationToCopyOntoAtlas.Height; row++)
+            {
+                int atlasStartingPoint = ((actualLocationToCopyOntoAtlas.Y + row) * atlasStride) + (actualLocationToCopyOntoAtlas.X * 4);
+                int ammountToCopy = textureData.Stride;
+                //https://stackoverflow.com/questions/1866236/add-offset-to-intptr
+                //Marshal.Copy(textureByte, 0, IntPtr.Add(atlasData.Scan0, atlasStartingPoint), ammountToCopy);
```

Comment:
Why did you comment out this line? Is it because you already did the copy above, or because the for loop following this is more effective/does the same thing?

Reply:
That looks to be leftover from when i was experimenting with how to copy sections of the texture data. im checking now

Reply:
ok i got it

so in this section we're calculating where to copy/overwrite the atlas image data, in stride segment, for each row of the texture image.

I had tried to use a Marshal copy one-liner, but it didn't end up working. So I settled for a for loop byte copy. Does the same thing.

Reply:
i improved the comments and remove the commented out code. Nice catch, thanks!